src/Event/RecipientGatewayEvent.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+namespace Drupal\sms\Event;
+
+use Symfony\Component\EventDispatcher\Event;
+use Drupal\sms\Entity\SmsGatewayInterface;
+
+/**
+ * Event fired to determine valid gateways for a recipient.
+ */
+class RecipientGatewayEvent extends Event {
+
+  /**
+   * The recipient phone number.
+   *
+   * @var string
+   */
+  protected $recipient;
+
+  /**
+   * An array of gateway doubles.
+   *
+   * @var array
+   *   The array of gateway/priority doubles where:
+   *     - Key 0: SmsGatewayInterface $gateway
+   *     - Key 1: integer $priority
+   */
+  protected $gateways = [];
+
+  /**
+   * Constructs the object.
+   *
+   * @param string $recipient
+   *   The recipient phone number.
+   */
+  public function __construct($recipient) {
+    $this->setRecipient($recipient);
+  }
+
+  /**
+   * Get the phone number for this event.
+   *
+   * @return string
+   *   The phone number for this event.
+   */
+  public function getRecipient() {
+    return $this->recipient;
+  }
+
+  /**
+   * Set the phone number for this event.
+   *
+   * @param string $recipient
+   *   The phone number for this event.
+   *
+   * @return $this
+   *   Return this event for chaining.
+   */
+  public function setRecipient($recipient) {
+    $this->recipient = $recipient;
+    return $this;
+  }
+
+  /**
+   * Get the gateways for this event.
+   *
+   * @return array
+   *   An array of doubles gateway/priority doubles.
+   */
+  public function getGateways() {
+    return $this->gateways;
+  }
+
+  /**
+   * Return gateways ordered by priority from highest to lowest.
+   *
+   * @return \Drupal\sms\Entity\SmsGatewayInterface[]
+   */
+  public function getGatewaysSorted() {
+    uasort($this->gateways, function($a, $b) {
```

Comment:
Looked at sms_rule_based#1. That looks good (once I can confirm I'm okay with the event approach).